InnoDB temporary tablespace encryption

gl-sergei opened the pull request:
Based on PR #1808
http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1093/

laurynas-biveinis commented:
Is general tablespace encryption (https://github.com/percona/percona-server/pull/1808) a prerequisite? If so, I'll get that reviewed thoroughly first so it can be merged and this PR reduced

gl-sergei commented:
Rebased on 5.7 trunk http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1203/

laurynas-biveinis commented on mysql-test/suite/innodb/t/temp_table_encrypt.test:
```diff
@@ -0,0 +1,93 @@
+
+SET GLOBAL innodb_temp_tablespace_encrypt = ON;
+
+# make sure encryption cannot be be turned ON/OFF for session
```
"be be", although I'd just leave this part for a sys_vars test and remove it from here altogether

gl-sergei replied:
removed

laurynas-biveinis commented on mysql-test/suite/innodb/t/temp_table_encrypt.test:
```diff
@@ -0,0 +1,93 @@
+CREATE INDEX t10_b ON t10 (b) ALGORITHM=COPY;
+DROP INDEX t10_b ON t10 ALGORITHM=COPY;
+
+# make sure server survives restart
```
What does it mean for the "server to survive restart"?

gl-sergei replied:
changed comment to "test that server comes booting up successfully after shutdown"

laurynas-biveinis commented on mysql-test/suite/innodb/t/temp_table_encrypt.test:
```diff
@@ -0,0 +1,93 @@
+
+# and restart without keyring plugin
+let $restart_hide_args=1
+let $restart_parameters = restart: ;
```
consistent spacing around = 
resetting restart_parameters to empty has the same effect as setting it to "restart:"

gl-sergei replied:
fixed

laurynas-biveinis commented on storage/innobase/fil/fil0fil.cc:
```diff
@@ -5470,18 +5470,33 @@ fil_io_set_encryption(
-					space->encryption_iv);
-		req_type.encryption_algorithm(Encryption::AES);
-	} else {
+	/* Don't encrypt page 0 of all tablespaces except redo log
```
"Don't encrypt page 0 of any tablespace except if it is the redo log tablespace" ?

gl-sergei replied:
changed to "Don't encrypt page 0 of all tablespaces except, don't encrypt redo log and system tablespace."

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -21349,6 +21375,11 @@ static MYSQL_SYSVAR_STR(temp_data_file_path, innobase_temp_data_file_path,
 
+static MYSQL_SYSVAR_BOOL(temp_tablespace_encrypt, srv_tmp_tablespace_encrypt,
+  PLUGIN_VAR_OPCMDARG,
+  "Enable or disable Encrypt of temporary tablespace.",
```
s/Encrypt/encryption

gl-sergei replied:
fixed

laurynas-biveinis commented on storage/innobase/srv/srv0srv.cc:
```diff
@@ -2668,6 +2671,55 @@ srv_master_do_shutdown_tasks(
+	}
+
+	if (is_system_tablespace(srv_tmp_space.space_id())) {
+		/* there is no separate tablespace for temorary tables */
```
"temporary"

gl-sergei replied:
fixed

percona-ysorokin commented on storage/innobase/fil/fil0fil.cc:
```diff
@@ -7353,9 +7369,9 @@ fil_set_encryption(
 	byte*			iv)
 {
-	ut_ad(!is_system_or_undo_tablespace(space_id));
+	ut_ad(!is_shared_system_tablespace(space_id));
```
Both assertion and subsequent `if(...) return` with the same condition

gl-sergei replied:
Yep. And it was so in upstream actually. It means in debug build server will crash and in release build it will refuse to create table.

gl-sergei commented on storage/innobase/srv/srv0srv.cc:
```diff
@@ -2749,6 +2801,9 @@ DECLARE_THREAD(srv_master_thread)(
 		}
 	}
 
+	/* Enable temorary tablespace encryption if set */
```
s/temorary/temporary/

gl-sergei replied:
fixed

gl-sergei commented:
All comments fixed and replied. Param build started, will update with the link once it completes.

laurynas-biveinis commented:
@gl-sergei, has the last Jenkins run completed?

gl-sergei commented:
@laurynas-biveinis sorry, forgot to paste the link. http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1212/

laurynas-biveinis commented on mysql-test/suite/innodb/t/temp_table_encrypt.test:
```diff
@@ -0,0 +1,87 @@
+CREATE INDEX t10_b ON t10 (b) ALGORITHM=COPY;
+DROP INDEX t10_b ON t10 ALGORITHM=COPY;
+
+# test that server comes booting up successfully after shutdown
```
"test restarting" would say the same?

laurynas-biveinis commented on mysql-test/suite/innodb/t/temp_table_encrypt.test:
```diff
@@ -0,0 +1,87 @@
+CREATE TEMPORARY TABLE t01 (a INT) ENGINE=InnoDB;
+INSERT INTO t01 VALUES (1), (2), (3);
+
+# and crash
```
Likewise "test crashing"

laurynas-biveinis commented on mysql-test/suite/innodb/t/temp_table_encrypt.test:
```diff
@@ -0,0 +1,87 @@
+DROP INDEX t10_b ON t10 ALGORITHM=COPY;
+
+# test that server comes booting up successfully after shutdown
+let $restart_hide_args= 1
```
No ";" at the end of line and no "--" at the start? Shouldn't this be a parse error?

gl-sergei replied:
strangely, it is not.. there is one more test `keyring_udf.test` with the same issue... I'll fix it. Actually not the only one... plenty of them ....

laurynas-biveinis commented on mysql-test/suite/innodb/t/temp_table_encrypt.test:
```diff
@@ -0,0 +1,87 @@
+INSERT INTO t01 VALUES (1), (2), (3);
+
+# and crash
+let $restart_hide_args=1
```
No ";" at the end of line and no "--" at the start? Shouldn't this be a parse error?

laurynas-biveinis commented on mysql-test/suite/innodb/t/temp_table_encrypt.test:
```diff
@@ -0,0 +1,87 @@
+INSERT INTO t01 VALUES (1), (2), (3);
+
+# and able to startup without keyring plugin
+let $restart_hide_args=1
```
Likewise

laurynas-biveinis commented on mysql-test/suite/innodb/t/temp_table_encrypt.test:
```diff
@@ -0,0 +1,87 @@
+CREATE TEMPORARY TABLE t01 (a INT) ENGINE=InnoDB;
+INSERT INTO t01 VALUES (1), (2), (3);
+
+let $restart_hide_args=1
```
Likewise

laurynas-biveinis commented on mysql-test/suite/sys_vars/t/innodb_temp_tablespace_encrypt_basic.test:
```diff
@@ -0,0 +1,64 @@
+# show that it's writable
+#
+set global innodb_temp_tablespace_encrypt=1;
+--sleep 2
```
This is unstable, same issue and fix as https://bugs.mysql.com/bug.php?id=87303

laurynas-biveinis commented on storage/innobase/fil/fil0fil.cc:
```diff
@@ -5470,18 +5470,33 @@ fil_io_set_encryption(
-					space->encryption_iv);
-		req_type.encryption_algorithm(Encryption::AES);
-	} else {
+	/* Don't encrypt page 0 of all tablespaces except, don't encrypt
```
Don't encrypt redo log and system tablespaces, for all the other types tablespaces, don't encrypt page 0
?

laurynas-biveinis commented on storage/innobase/fil/fil0fil.cc:
```diff
@@ -5470,18 +5470,33 @@ fil_io_set_encryption(
+		return;
+	}
+
+	/* For writting redo log, if encryption for redo log is disabled,
```
writing

gl-sergei replied:
I copied  this spelling from MySQL 8 I guess... Fixing

laurynas-biveinis commented on storage/innobase/fil/fil0fil.cc:
```diff
@@ -5470,18 +5470,33 @@ fil_io_set_encryption(
+		return;
 	}
+
+	/* For writting temporary tablespace, if encryption for temporary
```
writing

laurynas-biveinis commented on storage/innobase/fil/fil0fil.cc:
```diff
@@ -5470,18 +5470,33 @@ fil_io_set_encryption(
+	}
+
+	/* For writting redo log, if encryption for redo log is disabled,
+	skip set encryption. */
```
setting

laurynas-biveinis commented on storage/innobase/fil/fil0fil.cc:
```diff
@@ -5470,18 +5470,33 @@ fil_io_set_encryption(
 	}
+
+	/* For writting temporary tablespace, if encryption for temporary
+	tablespace is disabled, skip set encryption. */
```
setting

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -10977,18 +10977,44 @@ create_table_info_t::create_table_def()
-			/* Get a new table ID */
-			dict_table_assign_new_id(table, m_trx);
+				/* Get a new table ID */
+				dict_table_assign_new_id(table, m_trx);
```
I wonder why github shows the green bars here but not below. Tabs vs spaces?

gl-sergei replied:
No, just one more level of indentation added

laurynas-biveinis commented on storage/innobase/include/fsp0sysspace.h:
```diff
@@ -282,6 +282,17 @@ extern SysTablespace srv_sys_space;
+is_shared_system_tablespace(
+	ulint	id)
+{
+	return(id == srv_sys_space.space_id());
```
In theory the implementation should go fsp0sysspace.ic file, but I don't think we care about UNIV_NONINL (?) build

gl-sergei replied:
upstream `is_system_tablespace` and `is_system_or_undo_tablespace` are both in .h file

laurynas-biveinis commented on storage/innobase/include/os0file.h:
```diff
@@ -369,6 +369,12 @@ static const ulint ENCRYPTION_INFO_SIZE_V2 = (ENCRYPTION_MAGIC_SIZE \
 					 + 2 * sizeof(ulint));
 
+/** Default master key for bootstrap */
+static const char ENCRYPTION_DEFAULT_MASTER_KEY[] = "DefaultMasterKey";
```
This one is never used

laurynas-biveinis commented on storage/innobase/include/os0file.h:
```diff
@@ -369,6 +369,12 @@ static const ulint ENCRYPTION_INFO_SIZE_V2 = (ENCRYPTION_MAGIC_SIZE \
+static const char ENCRYPTION_DEFAULT_MASTER_KEY[] = "DefaultMasterKey";
+
+/** Default master key id for bootstrap */
+static const ulint ENCRYPTION_DEFAULT_MASTER_KEY_ID = 0;
```
This one is used in a single source file. Move there? Easy to change later as needed.

laurynas-biveinis commented on storage/innobase/srv/srv0srv.cc:
```diff
@@ -2668,6 +2671,55 @@ srv_master_do_shutdown_tasks(
+void
+srv_enable_temp_encryption_if_set()
+{
+	fil_space_t*	space;
```
Move to initialization

laurynas-biveinis commented on storage/innobase/srv/srv0srv.cc:
```diff
@@ -2668,6 +2671,55 @@ srv_master_do_shutdown_tasks(
+		return;
+	}
+
+	space = fil_space_get(srv_tmp_space.space_id());
```
Move to if

laurynas-biveinis commented on storage/innobase/srv/srv0srv.cc:
```diff
@@ -2668,6 +2671,55 @@ srv_master_do_shutdown_tasks(
+							  &mtr)) {
+				mtr_commit(&mtr);
+
+				ib::error() << "Can't set temporary"
```
Setting the variable and making the change are far from each other, how to communicate this to user? Who would not necessarily check the error log. Is ib::fatal() an option here?

gl-sergei replied:
agree, ib::fatal() makes sense to me

laurynas-biveinis commented on storage/innobase/srv/srv0start.cc:
```diff
@@ -1231,6 +1231,20 @@ srv_open_tmp_tablespace(
+							 Encryption::AES,
+							 NULL,
+							 NULL);
+				ut_ad(err == DB_SUCCESS);
```
If we are asserting here that fil_set_encryption is "impossible" (so that crashing a debug build instead of printing errors, aborting server startup is reasonable) then what about release build? Maybe ut_a here?

gl-sergei replied:
agree on that too

gl-sergei commented:
PR updated. Jenkins build: http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1325/

laurynas-biveinis commented on storage/innobase/srv/srv0srv.cc:
```diff
@@ -2756,6 +2807,9 @@ DECLARE_THREAD(srv_master_thread)(
 	}
 
+	/* Enable temporary tablespace encryption if set */
+	srv_enable_temp_encryption_if_set();
```
Why here - at the start of shutdown? Especially when the called function exits first thing in this srv_shutdown_state

gl-sergei replied:
hmm... I was pretty sure I was putting it inside of the loop :)

laurynas-biveinis replied:
Yet the current testcase succeeds - maybe need a new one where the server is started with off and switched to on dynamically?

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -10994,18 +10994,44 @@ create_table_info_t::create_table_def()
 			   && !Encryption::is_none(
-				   m_create_info->encrypt_type.str)) {
+				   m_create_info->encrypt_type.str)
+			   && !srv_tmp_tablespace_encrypt) {
```
Two loads in a row of srv_tmp_tablespace_encrypt can race with changing the variable value.

gl-sergei replied:
I don't think that changing of `srv_tmp_tablespace_encrypt` between loads here is a big deal. But now I think that instead of `srv_tmp_tablespace_encrypt` we must check encryption flag of `srv_tmp_space`.

gl-sergei commented:
Test case changed to start with encryption turned off (grep used to check that), then it turns encryption on and checks that by grep as well.
I put the code to enable encryption on existing tablespace into separate function in `fsp0fsp.cc`. Not sure if it would be better to have two functions - `fsp_` and `fil_` ones.
Reading through comments on my second PR, I think I should use `p_s` instead of `i_s` here as well.
Jenkins run is here: http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1364/

gl-sergei commented:
@laurynas-biveinis  Agree. But it is an issue with the test case. I was thinking about making the test require debug_sync for that reason. Do you think it is a good idea?

laurynas-biveinis commented:
No, it's a feature issue. User sets variable, how long has he wait until he's sure the setting has taken effect? Please check 8.0 redo log encryption how they solved this.

gl-sergei commented:
There is another possible issue which I need to investigate. When update to temp tablespace is done concurrently with CREATE TEMP TABLE. This actually may cause some damage.

gl-sergei commented:
Time window between variable update and it taking effect doesn't look like a big issue to me.

laurynas-biveinis commented:
If a sync bug can happen, it will, https://bugs.mysql.com/bug.php?id=87303. The testcases can sync by the InnoDB master thread loop count status variable

laurynas-biveinis commented:
On a second thought, please do review 8.0 behavior for the redo log encryption enabling. Syncing by thread loop count is fragile and if user needs confirmation that the feature is indeed enabled, it's the least intuitive thing to remember

gl-sergei commented:
@laurynas-biveinis  MySQL 8 doesn't care much about the time gape between variable being set and it taking effect. They seems to be OK with it to take into effect eventually or silently flipping back to OFF... Do you think changing our implementation to be synchronous is the right way to go?

laurynas-biveinis commented:
MySQL 8 does - the variable value changes only after it becomes effective, check that linked bug. It is not synchronous implementation (and ours does not have to be), it is the master thread which updates the variable value.

gl-sergei commented:
Variable changes immediately. It flips back to OFF in case of unsuccessful attempt to turn encryption on.

gl-sergei commented:
Looks like all comments are taken to attention. Behaviour is now exactly the same as in MySQL 8 when it turns ON encryption for redo/undo logs. Yes, there is still a time gap and it is impossible to know exact moment when change became effective. In case of error, variable becomes OFF and error message gets logged. Turning it OFF manually doesn't turn encryption off.

Added a test case creating temp table before encryption is enabled and destroying it afterwards. It also is trying to create a table when encryption is already enabled on FIL level, but haven't yet enabled at FSP level.

https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1384/

gl-sergei commented on mysql-test/suite/innodb/t/temp_table_encrypt.test:
```diff
@@ -0,0 +1,113 @@
+INSERT INTO t03 VALUES ('Curabitur laoreet, velit non interdum venenatis');
+
+# wait until pages flushed
+--let $wait_condition= SELECT COUNT(*) = 0 FROM performance_schema.global_status WHERE LOWER(variable_name) = 'innodb_buffer_pool_pages_dirty' AND variable_value > 0
```
Maybe `set global innodb_buf_flush_list_now = 1;` is better, but it requires debug build.

laurynas-biveinis replied:
Release/debug testing > debug testing, thus the current way is by necessity better

laurynas-biveinis replied:
How about setting innodb_max_dirty_pages_pct to 0?

gl-sergei replied:
Added innodb_max_dirty_pages_pct=0

laurynas-biveinis commented on mysql-test/suite/innodb/t/temp_table_encrypt.test:
```diff
@@ -0,0 +1,113 @@
+--source include/grep_pattern.inc
+
+# this table created in separate file per table tablespace
+CREATE TEMPORARY TABLE t04 (a INT) ENGINE=InnoDB ROW_FORMAT=COMPRESSED;
```
What is tested here?

gl-sergei replied:
This test intended to check that temp table created outside of shared temp tablespace has no issues. I added a check that it is encrypted as well.

laurynas-biveinis commented on mysql-test/suite/innodb/t/temp_table_encrypt.test:
```diff
@@ -0,0 +1,113 @@
+--error ER_ILLEGAL_HA_CREATE_OPTION
+CREATE TEMPORARY TABLE t05 (a INT) ENGINE=InnoDB ROW_FORMAT=COMPRESSED ENCRYPTION='y';
+
+CREATE TEMPORARY TABLE t06 (a INT) ENGINE=InnoDB ROW_FORMAT=COMPRESSED ENCRYPTION='n';
```
And here?

gl-sergei replied:
Basically I wanted to check that ENCRYPTION='y' is still not allowed for temp tables.

laurynas-biveinis commented on mysql-test/suite/innodb/t/temp_table_encrypt.test:
```diff
@@ -0,0 +1,113 @@
+INSERT INTO t10 (b) SELECT b FROM t10;
+INSERT INTO t10 (b) SELECT b FROM t10;
+INSERT INTO t10 (b) SELECT b FROM t10;
+INSERT INTO t10 (b) SELECT b FROM t10;
```
What is the purpose of making t10 big?

gl-sergei replied:
It needs to be big enough so that CREATE INDEX uses temp table

laurynas-biveinis commented on mysql-test/suite/innodb/t/temp_table_encrypt.test:
```diff
@@ -0,0 +1,113 @@
+INSERT INTO t01 VALUES (1), (2), (3);
+
+# test crashing
+let $restart_hide_args= 1;
```
Some MTR variable assignments done with space before =, some without

gl-sergei replied:
fixed

laurynas-biveinis commented on mysql-test/suite/innodb/t/temp_table_encrypt_debug.test:
```diff
@@ -0,0 +1,50 @@
+
+connection default;
+
+# unpause master thread
```
s/unpause/resume

gl-sergei replied:
fixed

laurynas-biveinis commented on storage/innobase/fsp/fsp0fsp.cc:
```diff
@@ -1050,6 +1045,53 @@ fsp_header_rotate_encryption(
+	ulint space_id)
+{
+	byte		encrypt_info[ENCRYPTION_INFO_SIZE_V2];
+	mtr_t		mtr;
```
Declare at first use

gl-sergei replied:
fixed

laurynas-biveinis commented on storage/innobase/fsp/fsp0fsp.cc:
```diff
@@ -1050,6 +1045,53 @@ fsp_header_rotate_encryption(
+	mtr_t		mtr;
+	fil_space_t*	space = fil_space_get(space_id);
+
+	const page_size_t	page_size(space->flags);
```
Likewise

gl-sergei replied:
fixed

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -21686,6 +21720,12 @@ static MYSQL_SYSVAR_BOOL(sync_debug, srv_sync_debug,
+static MYSQL_SYSVAR_ULONG(master_encrypt_debug,
+  srv_master_encrypt_debug,
+  PLUGIN_VAR_OPCMDARG,
+  "Pause in the middle of the enabling of temporary tablespace encryption.",
```
Could you please describe 0, 1, 2 values here to make it easier to use for any future testcases?

gl-sergei replied:
done

laurynas-biveinis commented on storage/innobase/include/fsp0sysspace.h:
```diff
@@ -282,6 +282,17 @@ extern SysTablespace srv_sys_space;
+@param[in]	id	Space ID to check
+@return true if id is a system tablespace, false if not. */
+UNIV_INLINE
+bool
```
warn_unused_result?

gl-sergei replied:
added

laurynas-biveinis commented on storage/innobase/include/os0file.h:
```diff
@@ -514,6 +514,11 @@ struct Encryption {
 
+#ifndef UNIV_INNOCHECKSUM
+	/** Check if keyring plugin loaded. */
+	static bool check_keyring();
```
warn_unused_result

gl-sergei replied:
done

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -9657,6 +9657,49 @@ Encryption::decrypt(
+/** Check if keyring plugin loaded. */
+bool Encryption::check_keyring()
+{
+	char*	key_type = NULL;
```
Declare at first uses

gl-sergei replied:
done

laurynas-biveinis commented on storage/innobase/srv/srv0srv.cc:
```diff
@@ -2675,6 +2681,73 @@ srv_master_do_shutdown_tasks(
 	return(n_bytes_merged || n_tables_to_drop);
 }
 
+void
```
static, header comment

gl-sergei replied:
done

laurynas-biveinis commented on storage/innobase/srv/srv0srv.cc:
```diff
@@ -2675,6 +2681,73 @@ srv_master_do_shutdown_tasks(
+		return;
+	}
+
+	if (srv_tmp_tablespace_encrypt) {
```
if (!srv_tmp_tablespace_encrypt) return (and make it 1st or 2nd if in the function, looks the cheapest), then everything below has one level less indentation

gl-sergei replied:
done

laurynas-biveinis commented on storage/innobase/srv/srv0srv.cc:
```diff
@@ -2675,6 +2681,73 @@ srv_master_do_shutdown_tasks(
+		if (srv_read_only_mode) {
+			srv_tmp_tablespace_encrypt = false;
+			ib::error() << "Can't set temporary tablespace to be"
+				<< " encrypted in read-only mode.";
```
InnoDB read only mode, to tell apart from the core server read only mode

gl-sergei replied:
done

laurynas-biveinis commented on storage/innobase/srv/srv0srv.cc:
```diff
@@ -2675,6 +2681,73 @@ srv_master_do_shutdown_tasks(
+			return;
+		}
+
+		fil_space_t* space = fil_space_get(srv_tmp_space.space_id());
```
const ?

gl-sergei replied:
const doesn't work because of `space->flags |= FSP_FLAGS_MASK_ENCRYPTION;`

laurynas-biveinis commented on storage/innobase/srv/srv0srv.cc:
```diff
@@ -2675,6 +2681,73 @@ srv_master_do_shutdown_tasks(
+		}
+
+		/* Make sure the keyring is loaded. */
+		if (!Encryption::check_keyring()) {
```
I wonder if this is really required here - fsp_enable_encryption below would call fsp_header_fill_encryption_info and would check the same anyway? But maybe worse diagnostics would be produced?

gl-sergei replied:
Yes, mainly because fsp_header_fill_encryption_info does not tell what was the error.

laurynas-biveinis commented on storage/innobase/srv/srv0srv.cc:
```diff
@@ -2675,6 +2681,73 @@ srv_master_do_shutdown_tasks(
+
+		ut_ad(fsp_is_system_temporary(space->id));
+
+		if (!FSP_FLAGS_GET_ENCRYPTION(space->flags)) {
```
Is it possible for this condition to be false legally? Maybe it can be replaced by an assert?

gl-sergei replied:
Yes, it possible and basically it is condition to not update FSP headers each time.

laurynas-biveinis commented on storage/innobase/srv/srv0srv.cc:
```diff
@@ -2675,6 +2681,73 @@ srv_master_do_shutdown_tasks(
+
+			if (!fsp_enable_encryption(srv_tmp_space.space_id())) {
+				srv_tmp_tablespace_encrypt = false;
+				ib::warn() << "Can't set temporary"
```
One failure above is ib::error(), these two are ib::warn() - should they all be the same?

gl-sergei replied:
changed all to be ib:error

laurynas-biveinis commented on storage/innobase/srv/srv0srv.cc:
```diff
@@ -2675,6 +2681,73 @@ srv_master_do_shutdown_tasks(
+
+			space->flags |= FSP_FLAGS_MASK_ENCRYPTION;
+
+			if (fil_set_encryption(space->id,
```
Looking at what fil_set_encryption may return under what conditions, it looks like this check can be replaced by an assert?

gl-sergei replied:
replaced by assert

laurynas-biveinis commented on storage/innobase/srv/srv0start.cc:
```diff
@@ -1231,6 +1231,20 @@ srv_open_tmp_tablespace(
+
+				fil_space_t*		space;
+
+				space = fil_space_get(temp_space_id);
```
Merge with declaration, declare as const

gl-sergei replied:
cannot be const, but merged with declaration

laurynas-biveinis commented:
Please also do a test MTR run with the feature and its supporting keyring enabled by default, for any crashes/surprises

gl-sergei commented:
Also 
- added test with keyring_vault which revealed race condition in the test case and fixed that race condition
- enforced encryption for file per table TEMP tables

Jenkins: https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1529/

I ran innodb test suite with keyring_file and innotb-temp-table-encrypt=ON locally, had trouble making keyring_file to be preloaded in Jenkins

laurynas-biveinis commented on mysql-test/suite/innodb/include/temp_table_encrypt.inc:
```diff
@@ -0,0 +1,158 @@
+
+# test restarting
+let $restart_hide_args = 1;
+let $restart_parameters = $KEYRING_RESTART_PARAM --innodb-temp-tablespace-encrypt --innodb_max_dirty_pages_pct=0;
```
innodb_max_dirty_pages_pct not required in this and further restarts?

gl-sergei replied:
well, maybe it is not required, but it does no harm either :) I think for this test case it even useful as it makes actual encryption to happen with higher probability.

laurynas-biveinis commented on storage/innobase/srv/srv0srv.cc:
```diff
@@ -2675,6 +2681,77 @@ srv_master_do_shutdown_tasks(
+
+#ifdef UNIV_DEBUG
+		if (srv_master_encrypt_debug != 0) {
+			srv_master_encrypt_debug = 2;
```
How is this working without atomics? Not even volatile, which would also be wrong, but at least it would force loads and stores as written

gl-sergei replied:
There is a `os_thread_sleep` call in the loop which can potentially use and modify any global variable, compiler would be terribly wrong if it would optimize `srv_master_encrypt_debug = 2` away.

gl-sergei commented:
Should I rebase this PR before the next review round?

percona-ysorokin commented on mysql-test/suite/perfschema/t/show_sanity.test:
```diff
@@ -416,6 +416,7 @@ insert into test.sanity values
   ("JUNK: GLOBAL-ONLY", "I_S.SESSION_VARIABLES", "INNODB_CHANGE_BUFFERING"),
   ("JUNK: GLOBAL-ONLY", "I_S.SESSION_VARIABLES", "INNODB_CHANGE_BUFFERING_DEBUG"),
   ("JUNK: GLOBAL-ONLY", "I_S.SESSION_VARIABLES", "INNODB_CHANGE_BUFFER_MAX_SIZE"),
+  ("JUNK: GLOBAL-ONLY", "I_S.SESSION_VARIABLES", "INNODB_MASTER_ENCRYPT_DEBUG"),
```
Try to maintain alphabetical order here.

percona-ysorokin replied:
`INNODB_TEMP_TABLESPACE_ENCRYPT` should also be added to this list.

percona-ysorokin commented on storage/innobase/os/os0file.cc:
```diff
@@ -9657,6 +9657,48 @@ Encryption::decrypt(
+	strcpy(key_name, ENCRYPTION_DEFAULT_MASTER_KEY);
+
+	/* We call key ring API to generate master key here. */
+	int	my_ret = my_key_generate(key_name, "AES",
```
Checking `my_ret` result here?

laurynas-biveinis commented:
Yes, please rebase

gl-sergei commented:
Already did today, and addressed Yura's comments. Here is the Jenkins build https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1623/

laurynas-biveinis commented on mysql-test/suite/innodb/include/temp_table_encrypt.inc:
```diff
@@ -0,0 +1,158 @@
+INSERT INTO t01 VALUES ('Lorem ipsum dolor sit amet, consectetur adipiscing elit');
+
+# wait until pages flushed
+--let $wait_condition= SELECT COUNT(*) = 0 FROM performance_schema.global_status WHERE LOWER(variable_name) = 'innodb_buffer_pool_pages_dirty' AND variable_value > 0
```
Looks like a typo: isn't wait_condition supposed to be wait *termination* condition, i.e. value = 0 ?

gl-sergei replied:
Actual condition here is `COUNT(*) = 0`, i.e. there is no variable with name `innodb_buffer_pool_pages_dirty` and `value > 0`

gl-sergei replied:
I'll rewrite it to be more readable

laurynas-biveinis commented on mysql-test/suite/innodb/include/temp_table_encrypt.inc:
```diff
@@ -0,0 +1,158 @@
+INSERT INTO t04 VALUES ('Praesent tristique eros a tempus fringilla');
+
+# wait until pages flushed
+--let $wait_condition= SELECT COUNT(*) = 0 FROM performance_schema.global_status WHERE LOWER(variable_name) = 'innodb_buffer_pool_pages_dirty' AND variable_value > 0
```
Likewise

laurynas-biveinis commented on mysql-test/suite/innodb/include/temp_table_encrypt.inc:
```diff
@@ -0,0 +1,158 @@
+INSERT INTO t03 VALUES ('Curabitur laoreet, velit non interdum venenatis');
+
+# wait until pages flushed
+--let $wait_condition= SELECT COUNT(*) = 0 FROM performance_schema.global_status WHERE LOWER(variable_name) = 'innodb_buffer_pool_pages_dirty' AND variable_value > 0
```
Likewise

laurynas-biveinis commented on mysql-test/suite/innodb/include/temp_table_encrypt.inc:
```diff
@@ -0,0 +1,158 @@
+CREATE TEMPORARY TABLE t04 (a TEXT) ENGINE=InnoDB ROW_FORMAT=COMPRESSED;
+INSERT INTO t04 VALUES ('Praesent tristique eros a tempus fringilla');
+
+# wait until pages flushed
```
Likewise

laurynas-biveinis commented on mysql-test/suite/innodb/include/temp_table_encrypt.inc:
```diff
@@ -0,0 +1,158 @@
+
+SET GLOBAL innodb_temp_tablespace_encrypt = ON;
+
+# create large enought table inorted to make CREATE INDEX to use temporary table
```
"enought", "inorted"

laurynas-biveinis commented on storage/innobase/include/srv0srv.h:
```diff
@@ -292,6 +292,9 @@ extern my_bool	srv_undo_log_truncate;
 /** UNDO logs not redo logged, these logs reside in the temp tablespace.*/
 extern const ulong	srv_tmp_undo_logs;
 
+/** Enable or disable Encrypt of temporary tablespace.*/
```
s/Encrypt/encryption

laurynas-biveinis commented on storage/innobase/srv/srv0srv.cc:
```diff
@@ -2675,6 +2681,77 @@ srv_master_do_shutdown_tasks(
+
+	if (!FSP_FLAGS_GET_ENCRYPTION(space->flags)) {
+
+		if (srv_read_only_mode) {
```
Add a test?

gl-sergei replied:
I guess this branch is not reachable because there is `ut_ad(!srv_read_only_mode);` in master thread. Will replace it with assert.

gl-sergei commented:
- changed commit message to refer JIRA issue
- rewrote SELECT to be a bit shorter
- replaced if (srv_read_only_mode) with assert, checked that it is indeed not reachable
- grammar errors in comments

https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1659/
